File: 0433-minimum-genetic-mutation/0433-minimum-genetic-mutation.py
from collections import deque


class Solution:
    def minMutation(self, startGene: str, endGene: str, bank: List[str]) -> int:
        if endGene not in bank: return -1

        que = deque([(startGene, 0)])
        chk = dict()

        for seq in bank:
            chk[seq] = False

        while que:
            gene, step = que.popleft()
            if gene == endGene: return step

            for i in range(len(gene)):
                for base in ('A', 'T', 'G', 'C'):
                    if gene[i] != base:
                        nextGene = gene[:i] + base + gene[i + 1:]
                        if nextGene in chk and not chk[nextGene]:
                            que.append((nextGene, step + 1))
                            chk[nextGene] = True

        return -1

File: 0433-minimum-genetic-mutation/0433-minimum-genetic-mutation.java
class Solution {
    public int minMutation(String startGene, String endGene, String[] bank) {
        if (!Arrays.asList(bank).contains(endGene)) {
            return -1;
        }

        Queue<String> que = new LinkedList<>();
        Set<String> chk = new HashSet<>();
        char[] base = {'A', 'T', 'G', 'C'};
        List<String> vault = Arrays.asList(bank);

        int result = 0;

        chk.add(startGene);
        que.add(startGene);

        while (!que.isEmpty()) {
            for (int i = que.size(); i > 0; i--) {
                String now = que.poll();
                if (now.equals(endGene)) {
                    return result;
                }
                char[] chrs = now.toCharArray();

                for (int j = 0; j < 8; j++) {
                    char original = chrs[j];

                    for (char c: base) {
                        chrs[j] = c;
                        String nextGene = new String(chrs);
                        if (!chk.contains(nextGene) && vault.contains(nextGene)) {
                            chk.add(nextGene);
                            que.add(nextGene);
                        }
                    }

                    chrs[j] = original;
                }
            }
            result++;
        }
        return -1;
    }
}